src/java-interop/java-interop-logger.h

```diff
@@ -44,10 +44,20 @@ void log_error (LogCategories category, const char *format, ...);
 
 void log_fatal (LogCategories category, const char *format, ...);
 
-void log_info (LogCategories category, const char *format, ...);
+void log_info_nocheck (LogCategories category, const char *format, ...);
 
 void log_warn (LogCategories category, const char *format, ...);
 
-void log_debug (LogCategories category, const char *format, ...);
+void log_debug_nocheck (LogCategories category, const char *format, ...);
+
+#define DO_LOG(_level, _category_, ...)                                           \
+	do {                                                                      \
+		if ((log_categories && ((_category_))) != 0) {                    \
+			::log_ ## _level ## _nocheck ((_category_), __VA_ARGS__); \
+		}                                                                 \
+	} while (0)
+
+#define log_debug(_category_, ...) DO_LOG (debug, (_category_), __VA_ARGS__)
+#define log_info(_category_, ...) DO_LOG (info, (_category_), __VA_ARGS__)
```

Comment:
I think these should be implemented as functions, so that we stay ABI compatible?

Or do it the other way around and keep the old ones untouched and define new ones as `log_<category>_checked` defines?

Reply:
The whole point of them is to avoid processing the argument list, thus avoid spending time evaluating arguments only to discard the message *inside* the function (if `log_info` were a function and you had a message with 10 placeholders, then 10 arguments would have to be prepared - function calls, memory accesses etc - to pass them to `log_info` which would then quit immediately because the category is disabled). And keeping their names allows us to automatically apply savings all over the monodroid code.

Reply:
My only concern is the ABI break. But because we (XA) are probably only consumer of these, it might be OK to change it, to save us from updating the XA/monodroid to use the new defines.

@jonpryor Is that OK with you?

Reply:
I discussed that with @jonpryor, he was the one who suggested the approach :)

Reply:
ok :-)